### PR TITLE
e2e: Ensure artifacts are saved at test completion for shared fixture

### PR DESCRIPTION
**Pull request by marun**
Previously, the new sharable fixture was running artifact-collecting functions registered at setup but not those registered during test execution. When executed in parallel, a sub-test's modification of a shared fixture's set of artifact functions wasn't propagating to the parent test.

This change switches to registering artifact functions with t.Cleanup() to ensure that artifact collection is tied to the scope of
the registering test rather than the scope of the fixture lifetime. It also ensures that artifacts for subtests using shared fixture are saved in a path specific to the subtest.

### CODE REVIEW

File: test/e2e/framework/fixture.go [hunk 1]
```diff
@@ -43,29 +38,30 @@ func NewKCPFixture(cfgs ...KcpConfig) *KCPFixture {
 	}
 }
 
-func (f *KCPFixture) SetUp(t *testing.T) func() {
-	var err error
-	f.ArtifactDir, f.DataDir, err = ScratchDirs(t)
+func (f *KCPFixture) SetUp(t *testing.T) {
+	artifactDir, dataDir, err := ScratchDirs(t)
```
[the remark this answers is not in the document]
Comment (marun) [reply test/e2e/framework/fixture.go:42]: Done.

**Comment by marun** [comment]
Updated to ensure:

- artifact collection compatibility with tests executed with RunParallel. Previously servers started by RunParallel were being killed before functions registered with t.Cleanup were run.
- server name is included in artifact paths
- artifacts for tests executed in parallel use the same base path.

**Comment by marun** [comment]
Removed check for INPROCESS on cluster controller's t.Parallel invocation. Developers will be responsible for setting `-p=1` to limit parallelism during debugging.

**Comment by marun** [comment]
Rebased to pick up new authorizer test and update it not to call fixture teardown.

File: test/e2e/framework/test.go [hunk 2]
```diff
@@ -100,10 +107,14 @@ func RunParallel(top *testing.T, name string, f TestFunc, cfgs ...KcpConfig) {
 			runningServers[server.name] = server
 		}
 
+		// Use a context to communicate when the test-executing
+		// goroutine has finished.
+		testCompleteCtx, testCompleteCancel := context.WithCancel(context.Background())
```
Comment (ncdc) [inline test/e2e/framework/test.go:112]: Do you want the parent to be ctx from L84 or for this to be disassociated from that?

File: test/e2e/reconciler/cluster/controller_test.go [hunk 3]
```diff
@@ -102,13 +103,6 @@ func TestClusterController(t *testing.T) {
 					return true
 				}, wait.ForeverTestTimeout, time.Millisecond*100)
 
-				defer servers[sourceClusterName].Artifact(t, func() (runtime.Object, error) {
```
[the remark this answers is not in the document]
Comment (marun) [reply test/e2e/reconciler/cluster/controller_test.go:105]: Yes. As per the commit comment, you recently updated the test to remove the artifacts targeted for collection so collection was reporting an error.

[the remark this answers is not in the document]
Comment (ncdc) [reply test/e2e/reconciler/cluster/controller_test.go:105]: Heh, right you are 😄

**Review by ncdc — APPROVED** [review]

**Pull request merged**